### ISSUE TITLE
Aria2 windows randomly freeze with my aria2.p12 cert using encryption

### ISSUE THREAD

**Issue by jc3213** [aria2/aria2#767]
APP: Aria2-1.28 windows x64
OS: Windows 10 Enterprise LTSB 2016 14393.351 x64
CONF: https://github.com/jc3213/Misc/raw/767c3336a7bb49a1cb7f6e507656091f93d42d35/Aria2/aria2.conf

It happens randomly that `aria2c.exe` eat up 26~28% of My CPU (Core i5-4460) and the command line window get frozen. Then the Aria2 WebUI started to report that `Oh Snap! Could not connect to the aria2 RPC server. Will retry in 10 secs. You might want to check the connection settings by going to Settings > Connection Settings`

The aria2.p12 I use is generated by OpenSSL following the instruction from IBM http://www.ibm.com/support/knowledgecenter/en/SSWHYP_4.0.0/com.ibm.apimgmt.cmc.doc/task_apionprem_gernerate_self_signed_openSSL.html

Here's the screenshot
![aria2 freeze](https://cloud.githubusercontent.com/assets/8744936/19960364/4643de1c-a1e7-11e6-92b7-281e32265918.png)

I'm reading https://github.com/ziahamza/webui-aria2/issues/108 and tried to generated another aria2.p12 certification. Hope this will work

**Comment by jc3213** [comment]
You can try start and finished a http download, while BitTorrent downloading is in progress to reproduce the problem. (which means BT is downloading [I have 4 torrents downloading], you start a http download, and the http download finished, but BT is still downloading)

Even though, even you only go with BT downloads, it will still make `aria2c.exe` freezing

**Comment by tatsuhiro-t** [comment]
Can you confirm that without encryption, aria2 does not freeze?

**Comment by jc3213** [comment]
@tatsuhiro-t Yes, it seems aria2 won't freeze without encryption. Or at least my aria2.p12 will freeze.

I've read some manuals though, they all based on Linux systems, I just arranged some code and configurations, and generated some more .p12 files based on them. With the new .p12 files, I can't access Aria2 RPC via Aria2 WebUI or any other download manager.

I get no idea how could I use encryption with Aria2.

**Comment by tatsuhiro-t** [comment]
I successfully reproduced this issue with pkcs12 file and on Windows 10 64bit.  aria2 is v1.28.0 64bit binary.

```
2016-11-08 23:06:21.062066 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 4 buffered: 0
2016-11-08 23:06:21.065100 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 3596 buffered: 0
2016-11-08 23:06:21.067739 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 3596 buffered: 0
2016-11-08 23:06:22.064622 [DEBUG] [WinTLSSession.cc:432] WinTLS: Read request: 4096 buffered: 0
2016-11-08 23:06:22.080217 [INFO] [rpc_helper.cc:103] Executing RPC method system.multicall
2016-11-08 23:06:22.083612 [DEBUG] [WinTLSSession.cc:432] WinTLS: Read request: 4096 buffered: 0
2016-11-08 23:06:22.085617 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 4 buffered: 0
2016-11-08 23:06:22.088624 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 4 buffered: 0
2016-11-08 23:06:22.090629 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 3596 buffered: 0
2016-11-08 23:06:22.091632 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 3596 buffered: 0
2016-11-08 23:06:23.094939 [DEBUG] [WinTLSSession.cc:432] WinTLS: Read request: 4096 buffered: 0
2016-11-08 23:06:23.096912 [DEBUG] [WinTLSSession.cc:432] WinTLS: Read request: 4096 buffered: 0
2016-11-08 23:06:23.096912 [INFO] [rpc_helper.cc:103] Executing RPC method system.multicall
2016-11-08 23:06:23.098918 [DEBUG] [WinTLSSession.cc:432] WinTLS: Read request: 4096 buffered: 0
2016-11-08 23:06:23.099920 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 4 buffered: 0
2016-11-08 23:06:23.100928 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 4 buffered: 0
2016-11-08 23:06:23.100928 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 3596 buffered: 0
2016-11-08 23:06:23.101939 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 3596 buffered: 0
2016-11-08 23:06:23.148049 [DEBUG] [WinTLSSession.cc:432] WinTLS: Read request: 4096 buffered: 0
2016-11-08 23:06:23.149051 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 2 buffered: 0
2016-11-08 23:06:23.149051 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 2 buffered: 0
2016-11-08 23:06:23.149051 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 2 buffered: 0
2016-11-08 23:06:23.150055 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 2 buffered: 0
2016-11-08 23:06:23.150055 [DEBUG] [WebSocketSessionMan.cc:64] WebSocket session removed.
2016-11-08 23:06:23.150055 [DEBUG] [WinTLSSession.cc:190] WinTLS: Closing connection
2016-11-08 23:06:23.150055 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 31 buffered: 0
```

aria2 froze after the log.  No more additional logging.  Ctrl-C did not work.  Closing browser didn’t help.

@nmaier Do you have any insight about this?

**Comment by tatsuhiro-t** [comment]
I got another suspicious log.  This might not be related to this issue directly.

```
2016-11-08 23:02:13.370176 [DEBUG] [WinTLSSession.cc:493] WinTLS: Connection abruptly closed!
2016-11-08 23:02:13.370176 [DEBUG] [WinTLSSession.cc:190] WinTLS: Closing connection
2016-11-08 23:02:13.370176 [DEBUG] [WinTLSSession.cc:284] WinTLS: Write request: 31 buffered: 0
2016-11-08 23:02:13.370176 [DEBUG] [WinTLSSession.cc:422] WinTLS: Write result: 31 buffered: 0
2016-11-08 23:02:13.370176 [DEBUG] [WinTLSSession.cc:242] WinTLS: Closed Connection
2016-11-08 23:02:13.370176 [INFO] [HttpServerCommand.cc:275] CUID#23 - Error occurred while reading HTTP request
Exception: [HttpServer.cc:180] errorCode=1 Got EOF from peer.
2016-11-08 23:02:13.370176 [DEBUG] [WinTLSSession.cc:185] WinTLS: Cannot close connection
```

**Comment by tatsuhiro-t** [comment]
I fixed this issue with commit d289dc1108324ff5589ab103ef0863faaa3c5de1

**Comment by nmaier** [comment]
yeah, the wonders of messing up signedness, good catch @tatsuhiro-t
